atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/InvalidationRunner.java

```diff
@@ -0,0 +1,205 @@
+        DatabaseMetaData metaData = connection.getMetaData();
+        ResultSet res = metaData.getTables(null, null, prefixedTimestampTableName(), null);
+
+        Preconditions.checkState(res.next(), "We are in the process of invalidating the "
+                + "InDbTimestampBoundStore but the data table does not exist. "
+                + "We should never reach here. Please contact support.");
+
+        return ImmutableLimits.builder()
+                .upperLimit(getColumnStatus(LAST_ALLOCATED, connection))
+                .legacyUpperLimit(getColumnStatus(LEGACY_LAST_ALLOCATED, connection))
+                .build();
+    }
+
+    private Optional<ColumnStatus> getColumnStatus(String colName, Connection connection) throws SQLException {
+        if (!PhysicalBoundStoreDatabaseUtils.hasColumn(connection, prefixedTimestampTableName(), colName)) {
+            return Optional.empty();
+        }
+
+        String sql = String.format("SELECT %s FROM %s FOR UPDATE", colName, prefixedTimestampTableName());
+        QueryRunner run = new QueryRunner();
+        return run.query(connection, sql, rs -> {
+            if (rs.next()) {
+                return ColumnStatus.columnStatusWithValue(rs.getLong(colName));
+            }
+            return ColumnStatus.columnStatusWithoutValue();
+        });
+    }
+
+    private TableStatus checkTableStatus(Limits limits) {
+        TableStatus status = getTableStatus(limits);
+
+        Preconditions.checkState(status != TableStatus.BOTH_COLUMNS,
+                "We detected the table has been poisoned but last_allocated column still exists."
+                        + "This is unexpected. Please contact support.");
+        return status;
+    }
+
+    private TableStatus getTableStatus(Limits limits) {
+        boolean upperLimitExists = limits.upperLimit().isPresent();
+        boolean legacyUpperLimitExists = limits.legacyUpperLimit().isPresent();
+
+        if (upperLimitExists) {
+            return legacyUpperLimitExists ? TableStatus.BOTH_COLUMNS : TableStatus.HEALTHY;
+        }
+        return legacyUpperLimitExists ? TableStatus.POISONED : TableStatus.NO_DATA; // no data in table
+    }
+
+    @Value.Immutable
+    interface Limits {
+        Optional<ColumnStatus> upperLimit();
+        Optional<ColumnStatus> legacyUpperLimit();
+    }
+
+    @Value.Immutable
+    interface ColumnStatus {
+        @Value.Default
+        default long value() {
+            return AtlasDbFactory.NO_OP_FAST_FORWARD_TIMESTAMP;
+        }
+
+        static Optional<ColumnStatus> columnStatusWithValue(long value) {
+            return Optional.of(ImmutableColumnStatus.builder().value(value).build());
+        }
+
+        static Optional<ColumnStatus> columnStatusWithoutValue() {
+            return Optional.of(ImmutableColumnStatus.builder().build());
+        }
+    }
+
+    private String prefixedTimestampTableName() {
+        return tablePrefix + timestampTable.getQualifiedName();
+    }
```

Comment:
nit: I'd put these above the interfaces/enums here; typically inner classes come at the end of the file.